### PR TITLE
LP conclusion page - Refs #11048

### PR DESCRIPTION
**Pull request by jloguercio**

### CODE REVIEW

File: main/badge/issued.php [hunk 1]
```diff
@@ -59,6 +59,15 @@
...
 
+// Open Graph Markup
+$htmlHeadXtra[] = "
+    <meta property='og:type' content='article' />
+    <meta property='og:title' content='".sprintf(get_lang('IHaveObtainedSkillXOnY'), $skillInfo['name'], api_get_setting('siteName'))."' />
+    <meta property='og:url' content='".api_get_path(WEB_PATH)."badge/".$skillId."/user/".$userId."' />
+    <meta property='og:description' content='".$skillInfo['description']."' />
+    <meta property='og:image' content='".$skillInfo['badge_image']."' />
+";
+
```
Comment (ywarnier) [inline main/badge/issued.php]: Está bien pero deberíamos ver si no entra en conflicto con la funcionalidad de Twitter Card (ver página de administración)
Maybe this enters in conflict with the Twitter Card settings on the admin page.

File: main/gradebook/lib/be/category.class.php [hunk 2]
```diff
@@ -1950,7 +1950,7 @@ public static function register_user_certificate($category_id, $user_id)
         $userHasSkills = false;
 
         if ($skillToolEnabled) {
-            if (!$category->getGenerateCertificates()) {
+            if ($category->getGenerateCertificates()) {
```
Comment (ywarnier) [inline main/gradebook/lib/be/category.class.php]: There's an error in the name of this function. I'll review that with you.

File: main/newscorm/final_item_template/template.html [hunk 3]
```diff
@@ -0,0 +1,5 @@
+<div>
+    Congratulations! You have finished this learning path
```
Comment (ywarnier) [inline main/newscorm/final_item_template/template.html]: This must be translated as well!

File: main/newscorm/learnpath.class.php [hunk 4]
```diff
@@ -5598,6 +5598,14 @@ public function return_new_tree($update_audio = 'false', $drop_element_here = fa
                 );
             }
 
+            // Detect if type is FINAL_ITEM to set path_id to SESSION
+            if ($arrLP[$i]['item_type'] == TOOL_FINAL_ITEM) {
```
Comment (ywarnier) [inline main/newscorm/learnpath.class.php]: It would be better to name it TOOL_FINAL_LP_ITEM

File: main/newscorm/learnpath.class.php [hunk 5]
```diff
@@ -5648,13 +5660,15 @@ public function return_new_tree($update_audio = 'false', $drop_element_here = fa
 
             if ($is_allowed_to_edit) {
                 if (!$update_audio || $update_audio <> 'true') {
-                    $move_icon .= '<a class="moved" href="#">';
-                    $move_icon .= Display::return_icon('move_everywhere.png', get_lang('Move'), array(), ICON_SIZE_TINY);
-                    $move_icon .= '</a>';
+                    if ($arrLP[$i]['item_type'] !== TOOL_FINAL_ITEM) {
```
Comment (ywarnier) [inline main/newscorm/learnpath.class.php]: This requires some code documentation: "// If it's a final item, we don't want the user to move it around, so hide the option".
Something like that.

File: main/newscorm/learnpath.class.php [hunk 6]
```diff
@@ -6306,6 +6318,22 @@ public function display_edit_item($item_id)
...
+                    $_SESSION['finalItem'] = true;
+                    $tbl_doc = Database :: get_course_table(TABLE_DOCUMENT);
+                    $sql = "SELECT lp.*, doc.path as dir
+                            FROM " . $tbl_lp_item . " as lp
+                            LEFT JOIN " . $tbl_doc . " as doc
+                            ON doc.id = lp.path
+                            WHERE
+                                lp.c_id = $course_id AND
+                                doc.c_id = $course_id AND
+                                lp.id = " . intval($item_id);
```
Comment (ywarnier) [inline main/newscorm/learnpath.class.php]: As often as possible, this should use the Doctrine syntax.

File: main/newscorm/learnpath.class.php [hunk 7]
```diff
@@ -6364,20 +6392,23 @@ public function display_resources()
         // Get al the forums.
         $forums = $this->get_forums(null, $course_code);
 
+        $finish = $this->getFinalItemForm();
```
Comment (ywarnier) [inline main/newscorm/learnpath.class.php]: Needs a little documentation here...

File: main/newscorm/learnpath.class.php [hunk 8]
```diff
@@ -10691,6 +10723,109 @@ public function createForum($forumCategoryId)
         return $forumId;
     }
 
+    private function getFinalItem()
```
Comment (ywarnier) [inline main/newscorm/learnpath.class.php]: This requires a PHPDoc block. The name of the function is too generic, but even if it wasn't, every function needs to be documented.

File: main/newscorm/learnpath.class.php [hunk 9]
```diff
@@ -10691,6 +10723,109 @@ public function createForum($forumCategoryId)
...
+        foreach ($this->items as $item) {
+            if ($item->type !== 'final_item') {
+                continue;
+            }
+
+            return $item;
+        }
+    }
+
+    private function getFinalItemTemplate()
```
Comment (ywarnier) [inline main/newscorm/learnpath.class.php]: Missing PHPDoc block

File: main/newscorm/learnpath.class.php [hunk 10]
```diff
@@ -10691,6 +10723,109 @@ public function createForum($forumCategoryId)
...
+    {
+        return file_get_contents(api_get_path(SYS_CODE_PATH) . 'newscorm/final_item_template/template.html');
+    }
+
+    private function getSavedFinalItem()
```
Comment (ywarnier) [inline main/newscorm/learnpath.class.php]: Missing PHPDoc block

File: main/newscorm/learnpath.class.php [hunk 11]
```diff
@@ -10691,6 +10723,109 @@ public function createForum($forumCategoryId)
...
+        $doc = DocumentManager::get_document_data_by_id($finalItem->path, $this->cc);
+
+        return file_get_contents($doc['absolute_path']);
+    }
+
+    /**
+     * 
+     * @return html
+     */
+    public function getFinalItemForm()
```
Comment (ywarnier) [inline main/newscorm/learnpath.class.php]: Missing PHPDoc documentation

File: main/newscorm/lp_add_item.php [hunk 12]
```diff
@@ -277,20 +277,20 @@ function confirmation(name) {
...
             break;
         case 'thread':
             echo $learnPath->display_thread_form('add', 0, $_GET['thread_id']);
             break;
-        case 'link':
+        case TOOL_LINK:
             echo $learnPath->display_link_form('add', 0, $_GET['file']);
             break;
-        case 'student_publication':
+        case TOOL_STUDENTPUBLICATION:
```
Comment (ywarnier) [inline main/newscorm/lp_add_item.php]: :+1:

File: main/newscorm/lp_controller.php [hunk 13]
```diff
@@ -1385,6 +1388,29 @@ function(reponse){
             'lp_id' => $_SESSION['oLP']->lp_id
         ]));
         break;
+    case 'add_final_item':
+        var_dump($_POST, $_GET, $lp_found);
```
Comment (ywarnier) [inline main/newscorm/lp_controller.php]: ?

File: main/newscorm/lp_final_item.php [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+$_in_course = true;
+require_once '../inc/global.inc.php';
+$current_course_tool  = TOOL_GRADEBOOK;
+
+api_protect_course_script(true);
+
+$courseCode = api_get_course_id();
+$userId = api_get_user_id();
+$sessionId = api_get_session_id();
+$id = isset($_GET['id']) ? $_GET['id'] : 0;
```
Comment (ywarnier) [inline main/newscorm/lp_final_item.php]: should be filtered with intval(), at least

File: main/newscorm/lp_final_item.php [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+$lpId = isset($_GET['lp_id']) ? $_GET['lp_id'] : 0;
```
Comment (ywarnier) [inline main/newscorm/lp_final_item.php]: should be filtered with intval(), at least

File: main/newscorm/lp_final_item.php [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+    true,
+    $sessionId
+);
+
+$finalItemTemplate = file_get_contents($documentInfo['absolute_path']);
+
+$finalItemTemplate = str_replace('((certificate))', $downloadCertificateLink, $finalItemTemplate);
+$finalItemTemplate = str_replace('((skill))', $badgeLink, $finalItemTemplate);
+
+$tpl = new Template(null, false, false);
```
Comment (ywarnier) [inline main/newscorm/lp_final_item.php]: A small line of documentation explaining why "null, false, false" would be nice (but not mandatory)

File: main/newscorm/learnpath.class.php [hunk 17]
```diff
@@ -5598,6 +5598,14 @@ public function return_new_tree($update_audio = 'false', $drop_element_here = fa
...
             }
 
+            // Detect if type is FINAL_ITEM to set path_id to SESSION
+            if ($arrLP[$i]['item_type'] == TOOL_FINAL_ITEM) {
+                $_SESSION['pathItem'] = $arrLP[$i]['path'];
+            }
+
+
+
+
```
Comment (ywarnier) [inline main/newscorm/learnpath.class.php]: A lot of blank lines...

File: main/newscorm/final_item_template/template.html [hunk 18]
```diff
@@ -0,0 +1,5 @@
...
+</div>
+((certificate)) </br>
```
Comment (jmontoyaa) [inline main/newscorm/final_item_template/template.html]: No debería ser <br /> ? 

Porqué  no  usar  twig??

Comment (jloguercio) [inline main/newscorm/final_item_template/template.html]: Esto es solo una plantilla, es decir se usa como ejemplo.

Comment (jloguercio) [inline main/newscorm/final_item_template/template.html]: Esto es solo una plantilla, es decir se usa solo como ejemplo.

Comment (jloguercio) [inline main/newscorm/final_item_template/template.html]: Aunque ... creo que si podría usar twig ... :8ball:

Comment (ywarnier) [inline main/newscorm/final_item_template/template.html]: Igual la nota sobre <br /> es cierta :-) No lo había notado

File: main/badge/issued.php (hunk 1 above)
Comment (jloguercio) [inline main/badge/issued.php]: Fixed now , i validate the twitter:card to replace the description, title and url to the actual badge page if is in this page and if twitter:card is enable

File: main/newscorm/final_item_template/template.html (hunk 3 above)
Comment (ywarnier) [inline main/newscorm/final_item_template/template.html]: After reviewing what it involves, we'll leave it like that for now.

File: main/gradebook/lib/be/category.class.php (hunk 2 above)
Comment (ywarnier) [inline main/gradebook/lib/be/category.class.php]: I cancel that request. The function is actually there to tell *if* we have configured the certificates generation or not

**Pull request merged**